### PR TITLE
move writemime to Images

### PR DESCRIPTION
**Pull request by SimonDanisch**
goes along with JuliaIO/ImageMagick.jl#10

### CODE REVIEW

**Comment by timholy** [comment]
I'm not a fan of this:
- It ties Images back to ImageMagick again, and will make it much harder to experiment with other loaders/savers.
- With FileIO, Images doesn't worry about I/O at all, and writemime is an I/O method
- Currently Images doesn't load ImageMagick but ImageMagick loads Images; consequently, ImageMagick can implement all of the necessary functionality by calling methods in Images.

If you think this is important, can you provide more justification?

**Comment by SimonDanisch** [comment]
I'm not a big fan either :(
But we need to put the full signature in a package that is always loaded
before calling writemime, as it can't be loaded on demand.. So it can only
live in FileIO or Images. It shouldn't live in FileIO, because FileIO can't
fully qualify the signature without loading Images. So Images is the only
package left. I tried to tie it not too closely to ImageMagick and I can
further reduce it. We can call the `getblob/image2magickwand` `image2png`
and then let other libraries implement this function as well... We can also
try to be smarter with the loading of the library. Maybe with some logic
from FileIO?!

**Comment by timholy** [comment]
How about having Images define
```jl
Base.writemime(io::IO, ::MIME"image/png", img::AbstractImage; mapi=nothing, minpixels=10^4, maxpixels=10^6) = error("Please load an I/O package via `using IOPkg`, where `IOPkg` would be ImageMagick or QuartzImageIO")
```

`ImageMagick.jl` and `QuartzImageIO.jl` can define a method with the identical signature and it will overwrite the one in Images.

**Comment by lucasb-eyer** [comment]
Though you'd also want to add a hint that it should be loaded *after* `Images` in that message, right?

**Comment by timholy** [comment]
It always will be, because ImageMagick looks like this:
```jl
module ImageMagick
using Images
# stuff
function Base.writemime(...)
   # blah
end

end  # module ImageMagick
```

**Comment by lucasb-eyer** [comment]
Ah makes sense, nevermind me then.

**Comment by SimonDanisch** [comment]
Does `QuartzImageIO` actually support this functionality? Haven't seen anything to write png's to streams, but haven't looked closely :D

**Comment by rsrock** [comment]
Not yet, but I've looked into it and it should be easy to add.

(Once I finish writing this manuscript, that is...)

**Comment by timholy** [comment]
I always write code before I write manuscripts. What's your problem? :wink: Oh, wait, maybe I'm the one with the problem...

**Comment by timholy** [comment]
I guess the one concern is https://github.com/JuliaLang/julia/issues/265; people may have to restart their IJulia kernel when they hit this. Still, it's not like this is the only thing forcing that on users, and it's not 100% obvious this will be a problem---if the mime type is created in the same function that dispatches to `writemime`, then we should be saved by type-instability.

**Comment by SimonDanisch** [comment]
Haha :D 
Okay... I'm not too happy, but I get the feeling that there won't come much happiness out of this issue eitherway...
I'd like to keep this part in Images, though, as it seems to be shared ultimately:
```Julia
   assert2d(img)
    A = data(img)
    nc = ncolorelem(img)
    npix = length(A)/nc
    while npix > maxpixels
        A = restrict(A, coords_spatial(img))
        npix = length(A)/nc
    end
    if npix < minpixels
        fac = ceil(Int, sqrt(minpixels/npix))
        r = ones(Int, ndims(img))
        r[coords_spatial(img)] = fac
        A = repeat(A, inner=r)
   end
```
Maybe we can keep this part of writemime in images and then do this at the end:
```Julia
if applicable(_writemime, io, mime, raw_data) 
_writemime(io, mime, raw_data) # _writemime, because otherwise we get into conflict with Array{Uint8,3}
else
error("Please load an I/O package via `using IOPkg`, where `IOPkg` would be ImageMagick or QuartzImageIO")
end
end

**Comment by SimonDanisch** [comment]
(or rather do the check at the beginning to avoid useless computations)

**Comment by rsrock** [comment]
> I always write code before I write manuscripts. What's your problem? :wink: Oh, wait, maybe I'm the one with the problem...

I'm with you on that.  Unfortunately, no Julia at all in this manuscript, so I can't even kid myself.

**Comment by timholy** [comment]
Definitely fine to keep any common parts in Images. The only part I object to is Images loading ImageMagick.

**Comment by SimonDanisch** [comment]
Good... Yeah, all this dynamic loading is terrible and best restricted to
one place

**Comment by SimonDanisch** [comment]
This should be good to merge!
Fixing REQUIRE maybe better in another PR.

File: src/writemime.jl [hunk 1]
```diff
@@ -0,0 +1,56 @@
+###
+### writemime
+###
+# only mime writeable to PNG if 2D (used by IJulia for example)
+mimewritable(::MIME"image/svg+xml", img::AbstractImage)             = false
+mimewritable(::MIME"image/png", img::AbstractImage) 				= sdims(img) == 2 && timedim(img) == 0
```
Comment (vchuravy) [inline src/writemime.jl]: There are some whitespace issues here

**Comment by SimonDanisch** [comment]
Thanks @vchuravy!
If no one protests, I'm going to merge this soon :)

Best,
Simon

**Comment by timholy** [comment]
Sorry to be out of touch, I am traveling and short on time. By all means, go for it!

**Pull request merged**